File: pattern-diagnose/assets/examples/python/adapter/after/notification_service.py
# SPDX-License-Identifier: MIT
"""After: extracted Adapter — translation logic centralized in one place."""

from typing import Any, Protocol


# Third-party library (cannot modify)
class ThirdPartyLogger:
    def log_message(self, level: int, msg: str, context: dict[str, Any] | None = None) -> None:
        print(f"[{level}] {msg} {context or ''}")


# Our application's expected interface
class AppLogger(Protocol):
    def debug(self, message: str) -> None: ...
    def info(self, message: str) -> None: ...
    def error(self, message: str, err: Exception | None = None) -> None: ...


# Adapter: translates AppLogger calls into ThirdPartyLogger calls
class LoggerAdapter:
    def __init__(self, target: ThirdPartyLogger) -> None:
        self._target = target

    def debug(self, message: str) -> None:
        self._target.log_message(0, message)

    def info(self, message: str) -> None:
        self._target.log_message(1, message, {"source": "app"})

    def error(self, message: str, err: Exception | None = None) -> None:
        self._target.log_message(3, message, {
            "source": "app",
            "error": str(err) if err else None,
        })


# Client code uses AppLogger — no knowledge of ThirdPartyLogger
class NotificationService:
    def __init__(self, logger: AppLogger) -> None:
        self._logger = logger

    def send(self, message: str) -> None:
        self._logger.info(message)
        self._logger.debug(f"notification sent: {message}")

File: pattern-diagnose/assets/examples/typescript/adapter/after/notification-service.ts
// SPDX-License-Identifier: MIT
// After: extracted Adapter — translation logic centralized in one place

// Third-party library (cannot modify)
interface ThirdPartyLogger {
  logMessage(level: number, msg: string, context?: Record<string, unknown>): void;
}

// Our application's expected interface
interface AppLogger {
  debug(message: string): void;
  info(message: string): void;
  error(message: string, err?: Error): void;
}

// Adapter: translates AppLogger calls into ThirdPartyLogger calls
class LoggerAdapter implements AppLogger {
  constructor(private target: ThirdPartyLogger) {}

  debug(message: string): void {
    this.target.logMessage(0, message);
  }

  info(message: string): void {
    this.target.logMessage(1, message, { source: "app" });
  }

  error(message: string, err?: Error): void {
    this.target.logMessage(3, message, {
      source: "app",
      error: err?.message,
    });
  }
}

// Client code uses AppLogger — no knowledge of ThirdPartyLogger
class NotificationService {
  constructor(private logger: AppLogger) {}

  send(message: string): void {
    this.logger.info(message);
    this.logger.debug(`notification sent: ${message}`);
  }
}
